### ISSUE TITLE
XML-RPC problem when checking content_type header

### ISSUE THREAD

**Issue by ZantsuRocks** [dokuwiki/dokuwiki#4218]
### The problem

The XML-RPC remote API don't check correctly the content_type header in the request.
When content type don't have the EXACT "text/xml" or "application/xml" string we receive the "XML-RPC server accepts XML request only" error
Some times we send the charset with comma separated in the content_type.

### Version of DokuWiki

2024-02-06a "Kaos"

### PHP Version

8.1.6

### Webserver and version of webserver

Apache 2.0 Windows Server 2022

### Relevant logs and/or error messages

```text
Change lines 40 and 41 in the XmlRpcServe.php file


str_contains(strtolower($_SERVER['CONTENT_TYPE']),'text/xml') &&
str_contains(strtolower($_SERVER['CONTENT_TYPE']),'application/xml')
```
```

**Comment by mprins** [comment]
I don't think it is a bug if software follows the specification: http://xmlrpc.com/spec.md

required headers are:
- A User-Agent and Host must be specified.
- The Content-Type is text/xml.
- The Content-Length must be specified and must be correct.

**Comment by Cherrg** [comment]
Hi there, 
some background information why the suggestion of @splitbrain is a valid solution and why this is actually a bug:

The bug was introduced on Jun 9, 2023 with this commit: https://github.com/dokuwiki/dokuwiki/commit/cc7691adb9a9fb131cdc07a53e826dbcd58bf0ad
and breaks existing xml-rpc client implementations.

The Content-Type header should be part of the internet protocol standard. E.g.:
- [RFC 2616](https://datatracker.ietf.org/doc/html/rfc2616#section-14.17) -> (drafted 1999, obsolete)
- [RFC 7231](https://datatracker.ietf.org/doc/html/rfc7231#section-3.1.1.5) -> (proposed standard June 2014, obsolete)
- [RFC 9110](https://datatracker.ietf.org/doc/html/rfc9110#section-8.3) -> (current standard, since 2022)

Quote from [RFC 5987](https://datatracker.ietf.org/doc/rfc5987/#3.2.1) (obsolete) / [RFC 8187](https://datatracker.ietf.org/doc/rfc8187/#3.2.1)  - this is about the charset

```
Note: recipients should be prepared to handle encoding errors,
      such as malformed or incomplete percent escape sequences, or non-
      decodable octet sequences, in a robust manner.  This specification
      does not mandate any specific behavior, for instance, the
      following strategies are all acceptable:

      *  ignoring the parameter,

      *  stripping a non-decodable octet sequence,

      *  substituting a non-decodable octet sequence by a replacement
         character, such as the Unicode character U+FFFD (Replacement
         Character).
```

I remember seeing an RFC specification that contains several formats with and without spaces for the character set in the content type header. -> [rfc7231#section-3.1.1.1](https://datatracker.ietf.org/doc/html/rfc7231#section-3.1.1.1)
Also check https://www.w3.org/Protocols/rfc1341/4_Content-Type.html for the complete definition of the content-type header field.

The charset is an optional part of the Content-Type header. Please just ignore the charset, or check for utf-8 if dokuwiki assumes this charset. It is even possible to specify the requested header in an answer: [Accept-Charset](https://datatracker.ietf.org/doc/html/rfc9110#name-accept-charset).

Funnily enough, the responses that the xmlrpc-api produces contain the following content-type header: “text/xml; charset=utf-8", however, currently do not accept this content type as before.

Have a nice day.
